pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -245,9 +306,149 @@ func (r *ReconcileNodeMaintenance) stopNodeMaintenance(nodeName string) error {
 		return err
 	}
 
+	if err := r.invalidateLease(nodeName); err != nil {
+		return err
+    }
+
 	return nil
 }
 
+func (r *ReconcileNodeMaintenance) invalidateLease(nodeName string) error {
+		if r.isLeaseSupported {
+				log.Info("Lease object supported,  invalidating lease")
+
+				nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+				lease := &coordv1beta1.Lease{}
+
+				if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+					return err
+				}
+				lease.Spec.AcquireTime = nil
+				lease.Spec.LeaseDurationSeconds = nil;
+				lease.Spec.RenewTime = nil
+				lease.Spec.LeaseTransitions = nil
+
+				if err := r.client.Update(context.TODO(), lease); err != nil {
+					return err
+				}
+		}
+		return nil
+}
+
+func (r *ReconcileNodeMaintenance) obtainLease(node *corev1.Node, durationInSeconds int) (reconcile.Result, error) {
+
+	lease, needUpdate, err := r.createOrGetExistingLease(node, durationInSeconds)
+
+	if err != nil {
+		log.Errorf("MaintOn: failed to create or get existing lease error=%v", err)
+		return reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}, err
+	}
+
+	if needUpdate {
+
+		log.Info("MaintOn: update lease - lease duration change requested")
+
+		if _, err = r.updateLease(node, lease, durationInSeconds, TransitionNoChange, false); err != nil {
+		return reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileNodeMaintenance) createOrGetExistingLease(node *corev1.Node, durationInSeconds int) (*coordv1beta1.Lease, bool, error) {
+	nodeName := node.ObjectMeta.Name
+
+	holderIdentity := LeaseHolderIdentity
+
+	owner := &metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        node.ObjectMeta.UID,
+	}
```

Comment:
OwnerReference isn't mandatory is it? What is the purpose of the setting one here?

Reply:
I remember having seen it on an actual lease object. As far as I understand it's purpose is for cleaning up dependent objects https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/